precompiles/setup.go

```diff
@@ -1,29 +1,50 @@
 package precompiles
 
 import (
+	"sync"
+
+	ecommon "github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/vm"
 	"github.com/sei-protocol/sei-chain/precompiles/bank"
 	"github.com/sei-protocol/sei-chain/precompiles/common"
+	"github.com/sei-protocol/sei-chain/precompiles/wasmd"
 )
 
+var SetupMtx = &sync.Mutex{}
+
 func InitializePrecompiles(
 	evmKeeper common.EVMKeeper,
 	bankKeeper common.BankKeeper,
+	wasmdKeeper common.WasmdKeeper,
+	wasmdViewKeeper common.WasmdViewKeeper,
 ) error {
+	SetupMtx.Lock()
+	defer SetupMtx.Unlock()
 	bankp, err := bank.NewPrecompile(bankKeeper, evmKeeper)
 	if err != nil {
 		return err
 	}
-	vm.PrecompiledContractsHomestead[bankp.Address()] = bankp
-	vm.PrecompiledContractsByzantium[bankp.Address()] = bankp
-	vm.PrecompiledContractsIstanbul[bankp.Address()] = bankp
-	vm.PrecompiledContractsBerlin[bankp.Address()] = bankp
-	vm.PrecompiledContractsCancun[bankp.Address()] = bankp
-	vm.PrecompiledContractsBLS[bankp.Address()] = bankp
-	vm.PrecompiledAddressesHomestead = append(vm.PrecompiledAddressesHomestead, bankp.Address())
-	vm.PrecompiledAddressesByzantium = append(vm.PrecompiledAddressesByzantium, bankp.Address())
-	vm.PrecompiledAddressesIstanbul = append(vm.PrecompiledAddressesIstanbul, bankp.Address())
-	vm.PrecompiledAddressesBerlin = append(vm.PrecompiledAddressesBerlin, bankp.Address())
-	vm.PrecompiledAddressesCancun = append(vm.PrecompiledAddressesCancun, bankp.Address())
+	addPrecompileToVM(bankp, bankp.Address())
+	wasmdp, err := wasmd.NewPrecompile(evmKeeper, wasmdKeeper, wasmdViewKeeper)
+	if err != nil {
+		return err
+	}
+	addPrecompileToVM(wasmdp, wasmdp.Address())
 	return nil
 }
+
+// This function modifies global variable in `vm` module. It should only be called once
```

Comment:
should we follow a singleton pattern here to ensure it's only initialized once (or something like baseapp.sealed where it panics if called multiple times)?

Reply:
made it idempotent. Not panicking because it may be called multiple times in tests